### DISCUSSION TITLE
Possible to detect the tint of the users terminal?

### DISCUSSION THREAD

**Discussion by erwin** [lrstanley/bubbletint#1] [General]
I noticed lipgloss has an adaptive color method to choose a light vs a dark color...

Is there a way we can gather enough color information about the users terminal to make an educated guess about which tint they have set as active (in their actual terminal settings)?

**Comment by lrstanley** [comment]
I believe only determining the default background and foreground color is possible via how `lipgloss` determines that. I think the two common ways are via Operating System Commands (`OSC`'s), where an escape sequence is used to query that information, or if the terminal emulator supports specific environment variables (e.g. `COLORFGBG`).

Both can only be achieved before the app fully starts (as the stdout can't have other data flowing through it).

I think the next problem is that I suspect there isn't enough detail in those query responses to tell us a ballpark tint.

May poke this in the future.

**Comment by erwin** [comment]
I didn't realize that the `OSC` codes were used to check foreground/background color, that makes sense! Fascinating the way `OSC` codes are getting a new life recently. [osc52](https://github.com/ojroques/nvim-osc52) is a lifesaver over `ssh`...

There's an app called `xtermcontrol` that apparently has support for querying all of the 16 primary colors of the terminal, so that may be enough to figure out what color scheme is applied... We would need to run some tests to see how well it's supported in modern terminals though.

https://web.archive.org/web/20170503214750/http://thrysoee.dk/xtermcontrol/

```
--get-fg
    Report foreground color.
--get-bg
    Report background color.
--get-colorN
    Report N'th [0-15] color.
--get-highlight
    Report highlight color.
```

It was a bit confusing to me the sometimes the color came from my terminal color scheme and sometimes the developer was overriding the color, until I read...

https://jeffkreeftmeijer.com/vim-16-color/

I'm sure you're familiar with this, but may be helpful for other people that come to learn more about terminal colors as they use `bubbletint`.

> ### 8-bit and 24-bit ANSI color
> 
> Beyond the first sixteen named colors numbered 1 through 16, we arrive in absolute color territory. The color codes from 16 to 255 are used to print 8-bit colors, from which 232 to 255 are grayscale values from black to white.
> 
> prints “hotpink” in #205 (HEX: #f69, RGB: 255, 102, 153)
>
> ```
> $ echo -e "\033[38;5;205mhotpink\033[m"
> ```
> 
> Then, there’s 24-bit color, also named true color, which has 16.777.216 different color values.
> 
> prints “giant goldfish” in 24-bit orange (HEX: #F38630, RGB: 243, 134, 48)
> 
> ```
> $ echo -e "\033[38;2;243;134;48mgiant goldfish\033[0m"
> ```
> 
> Being absolute color values, they’re not configurable in the terminal emulator’s settings.

Also, not sure if you've seen `gamut`, but it also has some interesting complementary color generation features.

https://github.com/muesli/gamut
